autode/opt/coordinates/primitives.py

```diff
@@ -40,17 +126,56 @@ def derivative(
 
         -----------------------------------------------------------------------
         Arguments:
-            i: Cartesian index to take the derivative with respect to. [0-N),
-               for N atoms
 
-            component: Cartesian component (x, y, z) to take the derivative
-                       with respect to
+            x: Cartesian coordinate array of shape (N, )
 
-            x: Cartesian coordinates
+        Returns:
+            (np.ndarray): Derivative array of shape (N, )
+        """
+        _x = x.ravel()
+        res = self._evaluate(_x, deriv_order=1)
+        derivs = np.zeros_like(_x, dtype=float)
+        for i in range(_x.shape[0]):
+            dqdx_i = res.differentiate_wrt(str(i))
+            if dqdx_i is not None:
+                derivs[i] = dqdx_i
+
+        return derivs
+
+    def second_derivative(
+        self,
+        x: "CartesianCoordinates",
+    ) -> np.ndarray:
+        r"""
+        Calculate the second derivatives with respect to cartesian coordinates
+
+        .. math::
+
+            \frac{d^2 q}
+                  {d\boldsymbol{X}_{i, k}^2} {\Bigg\rvert}_{X=X0}
+
+        where :math:`q` is the primitive coordinate and :math:`\boldsymbol{X}`
+        are the cartesian coordinates.
+
+        -----------------------------------------------------------------------
+        Arguments:
+
+            x: Cartesian coordinate array of shape (N, )
 
         Returns:
-            (float): Derivative
+            (np.ndarray): Second derivative matrix of shape (N, N)
         """
+        _x = x.ravel()
+        x_n = _x.shape[0]
+        res = self._evaluate(_x, deriv_order=2)
+        derivs = np.zeros(shape=(x_n, x_n), dtype=float)
+        for i in range(x_n):
+            for j in range(x_n):
+                d2q_dx2_ij = res.differentiate_wrt(str(i), str(j))
+                if d2q_dx2_ij is not None:
+                    derivs[i, j] = d2q_dx2_ij
```

Comment:
@t-young31 Yeah but it wouldn't matter much I think, because `differentiate_wrt()` just reads off the derivative matrix hidden in `VectorHyperDual` class, so it is simpler code I think to just go through all of them. (the second derivative matrix should be symmetric)